File: src/main/java/innexgo/hours/SchoolService.java
/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class SchoolService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public School getBySchoolId(long schoolId) {
    String sql =
        "SELECT * FROM school WHERE school_id=?";
    RowMapper<School> rowMapper = new SchoolRowMapper();
    School school = jdbcTemplate.queryForObject(sql, rowMapper, schoolId);
    return school;
  }

  public long nextId() {
    String sql = "SELECT max(school_id) FROM school";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }

  public void add(School school) {
    school.schoolId = nextId();
    school.creationTime = System.currentTimeMillis();
    // Add school
    String sql = "INSERT INTO school values (?, ?, ?, ?)";
    jdbcTemplate.update(
        sql,
        school.schoolId,
        school.creationTime,
        school.creatorUserId,
        school.whole);
  }

  public boolean existsBySchoolId(long schoolId) {
    String sql = "SELECT count(*) FROM school WHERE school_id=?";
    long count = jdbcTemplate.queryForObject(sql, Long.class, schoolId);
    return count != 0;
  }

  public Stream<School> query( //
     Long schoolId, //
     Long creationTime, //
     Long minCreationTime, //
     Long maxCreationTime, //
     Long creatorUserId, //
     String name, //
     String partialName, //
     Boolean whole, //
     long offset, //
     long count) //
 {

    String sql=
      "SELECT s.* FROM school s"
        + " WHERE 1=1 "
        + (schoolId        == null ? "" : " AND s.school_id = " + schoolId)
        + (creatorUserId   == null ? "" : " AND s.creator_user_id = " + creatorUserId)
        + (creationTime    == null ? "" : " AND s.creation_time = " + creationTime)
        + (minCreationTime == null ? "" : " AND s.creation_time > " + minCreationTime)
        + (maxCreationTime == null ? "" : " AND s.creation_time < " + maxCreationTime)
        + (name            == null ? "" : " AND s.name = " + Utils.escape(name))
        + (partialName     == null ? "" : " AND s.name LIKE " + Utils.escape("%"+partialName+"%"))
        + (whole           == null ? "" : " AND s.whole = " + whole)
        + (" ORDER BY s.school_id")
        + (" LIMIT " + offset + ", " + count)
        + ";";

    RowMapper<School> rowMapper = new SchoolRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }
}

File: src/school_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use tokio_postgres::GenericClient;
use innexgo_hours_api::request;

impl From<tokio_postgres::row::Row> for School {
  // select * from school order only, otherwise it will fail
  fn from(row: tokio_postgres::row::Row) -> School {
    School {
      school_id: row.get("school_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      whole: row.get("whole"),
    }
  }
}

pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  whole: bool,
) -> Result<School, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let school_id = con
    .query_one(
      "INSERT INTO
       school(
           creation_time,
           creator_user_id,
           whole
       )
       VALUES($1, $2, $3)
       RETURNING school_id
      ",
      &[&creation_time, &creator_user_id, &whole],
    ).await?
    .get(0);

  // return school
  Ok(School {
    school_id,
    creation_time,
    creator_user_id,
    whole
  })
}

pub async fn get_by_school_id(
  con: &mut impl GenericClient,
  school_id: i64,
) -> Result<Option<School>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM school WHERE school_id=$1",
      &[&school_id],
    ).await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: request::SchoolViewProps,
) -> Result<Vec<School>, tokio_postgres::Error> {
  let results = con
    .query(
      "
        SELECT sc.* FROM school sc WHERE 1 = 1
        AND ($1::bigint[] IS NULL OR sc.school_id IN $1)
        AND ($2::bigint   IS NULL OR sc.creation_time >= $2)
        AND ($3::bigint   IS NULL OR sc.creation_time <= $3)
        AND ($4::bigint[] IS NULL OR sc.creator_user_id IN $4)
        AND ($5::bool     IS NULL OR sc.whole = $5)
        ORDER BY sc.school_id
      ",
      &[
        &props.school_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.whole,
      ],
    ).await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
